File: DataStructures/BinaryTree.js
class Node {
	constructor({ value, leftNode, rightNode }) {
		if (!value) {
			throw new Error('missing params in node creation')
		}
		this.value = value
		this.leftNode = leftNode
		this.rightNode = rightNode
	}
}

class BinaryTree {
	rootNode = null
	push(value) {
		if (this.rootNode === null) {
			this.rootNode = new Node({ value })
			return
		}

		var currentNodeToAnalyze = this.rootNode

		while (true) {
			if (currentNodeToAnalyze.value < value) {
				if (currentNodeToAnalyze.rightNode === undefined) {
					currentNodeToAnalyze.rightNode = new Node({ value })
					return
				}
				currentNodeToAnalyze = currentNodeToAnalyze.rightNode
				continue
			}
			if (currentNodeToAnalyze.value > value) {
				if (currentNodeToAnalyze.leftNode === undefined) {
					currentNodeToAnalyze.leftNode = new Node({ value })
					return
				}
				currentNodeToAnalyze = currentNodeToAnalyze.leftNode
				continue
			}

			return 'whoops this number already have been inserted'
		}
	}

	delete({ nodeValue, nodeToEvaluate }) {
		if (!nodeToEvaluate) {
			return 'number not found'
		}

		if (nodeToEvaluate.value === nodeValue) {
			nodeToEvaluate = null
			return 'removed'
		}

		if (
			nodeToEvaluate.rightNode &&
			nodeToEvaluate.rightNode.value === nodeValue
		) {
			nodeToEvaluate.rightNode = undefined
			return 'removed'
		}

		if (
			nodeToEvaluate.leftNode &&
			nodeToEvaluate.leftNode.value === nodeValue
		) {
			nodeToEvaluate.leftNode = undefined
			return 'removed'
		}

		if (nodeValue > nodeToEvaluate.value) {
			return this.delete({
				nodeValue,
				nodeToEvaluate: nodeToEvaluate.rightNode,
			})
		}
		if (nodeValue < nodeToEvaluate.value) {
			return this.delete({
				nodeValue,
				nodeToEvaluate: nodeToEvaluate.left,
			})
		}
		return 'got to end'
	}
}

const binaryTree = new BinaryTree()

binaryTree.push(1)
binaryTree.push(2)
binaryTree.push(3)
binaryTree.push(-1)

binaryTree.delete({ nodeValue: 3, nodeToEvaluate: binaryTree.rootNode })
binaryTree.delete({ nodeValue: 2, nodeToEvaluate: binaryTree.rootNode })
console.log(binaryTree.rootNode)
binaryTree.delete({ nodeValue: -1, nodeToEvaluate: binaryTree.rootNode })
console.log(binaryTree.rootNode)

File: DataStructures/BinaryTree.py
class Node:
    value = None
    rightNode = None
    leftNode = None

    def __init__(self, **nodeValues):
        self.value = nodeValues.get('value')
        self.rightNode = nodeValues.get('rightNode')
        self.leftNode = nodeValues.get('leftNode')

class BinaryTree:
    rootNode = None

    def push(self, value):
        if self.rootNode == None:
            self.rootNode = Node(value=value)
            return 'inserted'

        currentNode = self.rootNode

        while(True):
            if currentNode.value < value:
                if not currentNode.rightNode:
                    currentNode.rightNode = Node(value=value)
                    return 'inserted'
                currentNode = currentNode.rightNode
                continue
            if currentNode.value > value:
                if not currentNode.leftNode:
                    currentNode.leftNode = Node(value=value)
                    return 'inserted'
                currentNode = currentNode.leftNode
                continue

            return 'this number was already inserted'

    def delete(self, currentNode, valueToDelete):
        if not currentNode:
            return 'number not found'
        
        if currentNode.rightNode and currentNode.rightNode.value == valueToDelete:
            currentNode.rightNode = None
            return 'removed'
        if currentNode.leftNode and currentNode.leftNode.value == valueToDelete:
            currentNode.leftNode = None
            return 'removed'

        if currentNode.value < valueToDelete:
            return self.delete(currentNode.rightNode, valueToDelete)
        if currentNode.value < valueToDelete:
            return self.delete(currentNode.leftNode, valueToDelete)


binaryTree = BinaryTree()

print(binaryTree.push(1))
print(binaryTree.push(2))
print(binaryTree.push(3))
print(binaryTree.delete(binaryTree.rootNode, 2))
print(binaryTree.delete(binaryTree.rootNode, 3))